File: qpid/cpp/src/qpid/store/ms-clfs/Transaction.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

#include "Transaction.h"
#include "Messages.h"

namespace qpid {
namespace store {
namespace ms_clfs {

Transaction::~Transaction()
{
    // Transactions that are recovered then found to be deleted get destroyed
    // but need not be logged.
    if (state != TRANS_DELETED)
        log->deleteTransaction(id);
}

void
Transaction::enroll(uint64_t msgId)
{
    qpid::sys::ScopedWlock<qpid::sys::RWlock> l(enrollLock);
    enrolledMessages.push_back(msgId);
}

void
Transaction::unenroll(uint64_t msgId)
{
    qpid::sys::ScopedWlock<qpid::sys::RWlock> l(enrollLock);
    for (std::vector<uint64_t>::iterator i = enrolledMessages.begin();
         i < enrolledMessages.end();
         ++i) {
        if (*i == msgId) {
            enrolledMessages.erase(i);
            break;
        }
    }
}

void
Transaction::abort(Messages& messages)
{
    log->recordAbort(id);
    for (size_t i = 0; i < enrolledMessages.size(); ++i)
        messages.abort(enrolledMessages[i], shared_from_this());
    state = TRANS_ABORTED;
}

void
Transaction::commit(Messages& messages)
{
    log->recordCommit(id);
    for (size_t i = 0; i < enrolledMessages.size(); ++i)
        messages.commit(enrolledMessages[i], shared_from_this());
    state = TRANS_COMMITTED;
}

void
TPCTransaction::prepare(void)
{
    log->recordPrepare(id);
    state = TRANS_PREPARED;
}

}}}  // namespace qpid::store::ms_clfs

File: qpid/java/broker-core/src/main/java/org/apache/qpid/server/store/Transaction.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.server.store;

import org.apache.qpid.server.message.EnqueueableMessage;
import org.apache.qpid.server.util.FutureResult;

public interface Transaction
{

    /**
     * Places a message onto a specified queue, in a given transactional context.
     *
     *
     *  @param queue     The queue to place the message on.
     * @param message
     */
    MessageEnqueueRecord enqueueMessage(TransactionLogResource queue, EnqueueableMessage message);

    void dequeueMessage(MessageEnqueueRecord enqueueRecord);

    /**
     * Commits all operations performed within a given transactional context.
     *
     */
    void commitTran();

    /**
     * Commits all operations performed within a given transactional context.
     *
     */
    FutureResult commitTranAsync();

    /**
     * Abandons all operations performed within a given transactional context.
     *
     */
    void abortTran();


    interface EnqueueRecord
    {
        TransactionLogResource getResource();
        EnqueueableMessage getMessage();
    }

    interface DequeueRecord
    {
        MessageEnqueueRecord getEnqueueRecord();
    }

    void removeXid(StoredXidRecord record);


    StoredXidRecord recordXid(long format, byte[] globalId, byte[] branchId, EnqueueRecord[] enqueues,
                   DequeueRecord[] dequeues);

    interface StoredXidRecord
    {
        long getFormat();
        byte[] getGlobalId();
        byte[] getBranchId();

    }
}
